### PR TITLE
Add Sentinel readonly replica support

### PR DESCRIPTION
**Pull request by smf8**
Hi, This is my suggested implementation for a read-only replica client (#132)  for **sentinel** Redis setup. Please share your thoughts on this implementation. If it's OK with you I can write the unit tests to cover what's changed.

### CODE REVIEW

File: sentinel.go [hunk 1]
```diff
@@ -289,19 +305,75 @@ func (c *sentinelClient) _refresh() (err error) {
...
+	// otherwise send master as addrress
+	m, err := resp[1].AsStrSlice()
+	if err != nil {
+		return "", nil, err
+	}
+	return fmt.Sprintf("%s:%s", m[0], m[1]), sentinels, nil
+}
+
+// watchEvents will help subscribing to channel ONLY ONCE.
+func (c *sentinelClient) watchEvents(ctx context.Context, cc conn) {
```
Comment (smf8) [inline sentinel.go:371]: This is intentional because in the previous implementation, every time a message is received, eventually it will cause another launch of `cc.Receive(...)` which then will lead to multiple operations for a single event.

Comment (rueian) [reply sentinel.go:371]: Hi @smf8,

Since you mentioned another issue here, could we deal with it first and in a separate PR?

I just looked back into the `refresh` and `listWatch`, and I think they would only cause duplicated `cc.Receive(...)` when `_switchMaster` failed and we can quickly fix the duplication by always closing the previous `sConn` in the `_refresh`.

Comment (rueian) [reply sentinel.go:371]: Sorry, we need to send UNSUBSCRIBE first before we close the previous `sConn` so that the previous `cc.Receive(...)` can exit without error.

**Comment by rueian** [comment]
Hi @smf8,

Correct me if I am wrong. So the `ReadOnlyReplica` will restrict the client to only connect to replicas. How can this be useful?

**Comment by smf8** [comment]
Yes exactly.
This can provide the option for clients to define separate instances of `rueidis.Client` for read/write operations. 
I thought about using `[]conn` in client but It will introduce many complexities with not much gain in my opinion.

File: sentinel.go (hunk 1 above)
Comment (smf8) [reply sentinel.go:371]: How about only sending an `UNSUBSCRIBE`  command before starting `cc.Receive` in the main goroutine, if there wasn't any previous subscription it will execute without doing anything. So there won't be any need to close the connection.

Comment (rueian) [reply sentinel.go:371]: Yes, good idea.

Comment (smf8) [reply sentinel.go:371]: #249

File: internal/cmds/gen_sentinel.go [hunk 2]
```diff
@@ -2,6 +2,26 @@
 
 package cmds
 
+type SentinelReplicas Completed
```
Comment (rueian) [inline internal/cmds/gen_sentinel.go:5]: These `Completed` types are generated by `go run gen.go` from https://github.com/redis/rueidis/blob/main/hack/cmds/commands_sentinel.json.

The JSON file should be updated as well.

File: rueidis.go [hunk 3]
```diff
@@ -135,6 +135,10 @@ type ClientOption struct {
 	AlwaysPipelining bool
 	// AlwaysRESP2 makes rueidis.Client always uses RESP2, otherwise it will try using RESP3 first.
 	AlwaysRESP2 bool
+
+	// ReadOnlyReplica indicates that this client will only try to connect to readonly replicas of redis setup.
+	// currently, it is only implemented for sentinel client
+	ReadOnlyReplica bool
```
Comment (rueian) [inline rueidis.go:141]: Instead of `ReadOnlyReplica`, I think `ReplicasOnly` would be better.

File: sentinel.go [hunk 4]
```diff
@@ -36,19 +38,21 @@ func newSentinelClient(opt *ClientOption, connFn connFn) (client *sentinelClient
...
+	sentinels    *list.List
+	mAddr        string
+	sAddr        string
+	sc           call
+	mu           sync.Mutex
+	stop         uint32
+	cmd          cmds.Builder
+	retry        bool
+	readonly     bool
+	watchCounter atomic.Int32
```
Comment (rueian) [inline sentinel.go:55]: We don't need the `watchCounter` now, right?

File: sentinel.go [hunk 5]
```diff
@@ -319,13 +339,28 @@ func (c *sentinelClient) listWatch(cc conn) (master string, sentinels []string,
...
 	}(cc)
 
-	resp := cc.DoMulti(ctx, sentinelsCMD, getMasterCMD)
+	commands := []Completed{
+		sentinelsCMD,
+		getMasterCMD,
+	}
+
+	if c.readonly {
+		commands = append(commands, replicasCMD)
```
Comment (rueian) [inline sentinel.go:360]: It is ok just always send the `replicasCMD`.

File: sentinel.go [hunk 6]
```diff
@@ -319,13 +339,28 @@ func (c *sentinelClient) listWatch(cc conn) (master string, sentinels []string,
 				if m[0] == "master" && m[1] == c.sOpt.Sentinel.MasterSet {
 					c.switchMasterRetry(fmt.Sprintf("%s:%s", m[2], m[3]))
 				}
+			case "+slave":
```
Comment (rueian) [inline sentinel.go:342]: Should we subscribe to `-slave` as well?

**Comment by rueian** [comment]
> Yes exactly. This can provide the option for clients to define separate instances of `rueidis.Client` for read/write operations.

Yes, indeed there might be some use case for reading only from replicas.

However, this PR can't actually work yet, right? Since we still need to set the underlying connection in the `READONLY` state, otherwise all requests will be rejected by replicas.

File: internal/cmds/gen_sentinel.go (hunk 2 above)
Comment (smf8) [reply internal/cmds/gen_sentinel.go:5]: Sorry, I missed that. Done.

File: sentinel.go (hunk 6 above)
Comment (smf8) [reply sentinel.go:342]: I couldn't find this channel in official sentinel documentation ([here](https://redis.io/docs/management/sentinel/#pubsub-messages)). Also removing a slave replica locally doesn't seem to trigger it. Although `+sdown` and `-sdown` is triggered when a replica (or master) is restarted and we can capture it with condition `m[0] == "slave"`.

Comment (rueian) [reply sentinel.go:342]: I am sorry about the misinformation. What about failover-end?

Comment (smf8) [reply sentinel.go]: I think the only condition that's not covered by current channels is the condition of `slave` shutting down. master restart/shutdowns are covered with `+switch-master` and `+reboot`. Also `sdown` status of replicas isn't reported by `failover-end` as I've tested. 

What do you think about the changes proposed in ab120af1ad7695448e6bc7e718f9a045820da152 and 37741dd8065a9601b4c33c5209ff70545d4b9bb4?

Feel free to check your other comments and check if they're resolved.

**Comment by smf8** [comment]
> However, this PR can't actually work yet, right

I think aside from some unit/integration tests to cover all edge cases. It's ready to use. Since this option is only for `sentinel` clients and with a sentinel setup, all slaves are started with the `replica-of` config enabled. There is no need to mark the connection as `READONLY`. 

Also since the user needs to **explicitly** mark the client as `ReplicaOnly`, receiving errors on write commands will be expected behavior.

File: sentinel.go (hunk 6 above)
Comment (rueian) [reply sentinel.go]: Monitoring `+sdown` and `-sdown` looks good to me.

Just one concern about checking the existence of `s-down-time`: Will it be clear once the replica is alive?

Besides that, I think the way you `delete response i` may panic: https://go.dev/play/p/zWeXCrTNqwJ

Comment (smf8) [reply sentinel.go]: Thanks for the precise review :)

Sorry, I've updated the method to avoid panic.

Also `s-down-time` seems to disappear once the replica is live and can reply to sentinel `PING`s. I couldn't find any documentation on it to confirm but I've tested it with multiple scenarios and it looks like to work that way.

File: sentinel.go [hunk 7]
```diff
@@ -335,6 +370,41 @@ func (c *sentinelClient) listWatch(cc conn) (master string, sentinels []string,
...
+		}
+
+		if len(eligibleReplicas) == 0 {
+			return "", nil, fmt.Errorf("not enough ready replicas")
+		}
+
+		// choose a replica randomly
+		randomReplicaIndex := rand.Intn(len(eligibleReplicas))
+		m, err := eligibleReplicas[randomReplicaIndex].AsStrMap()
+		if err != nil {
```
Comment (rueian) [inline sentinel.go:400]: We can skip the error checking here, since we have checked them above (if we also use AsStrMap above).

File: sentinel.go [hunk 8]
```diff
@@ -335,6 +370,41 @@ func (c *sentinelClient) listWatch(cc conn) (master string, sentinels []string,
...
+	if c.readonly {
+		replicaResponses, err := resp[1].ToArray()
+		if err != nil {
+			return "", nil, err
+		}
+
+		eligibleReplicas := make([]RedisMessage, 0, len(replicaResponses))
+		// eliminate replicas with s_down condition
+		for i := range replicaResponses {
+			replicaResponseMap, err := replicaResponses[i].AsMap()
```
Comment (rueian) [inline sentinel.go:384]: AsMap can be changed to AsStrMap to align with the below one.

**Comment by rueian** [comment]
Hey @smf8, I have tried this PR manually and it worked well.

So one thing left is about testing, could you try to make test coverage of this patch 100%?

**Comment by smf8** [comment]
That's awesome 🎉🎉. 

Sure, I'll try updating the tests to cover all the changes tomorrow. let me know if you can think of anything else for this patch.

File: sentinel.go [hunk 9]
```diff
@@ -319,13 +338,29 @@ func (c *sentinelClient) listWatch(cc conn) (master string, sentinels []string,
 				if m[0] == "master" && m[1] == c.sOpt.Sentinel.MasterSet {
 					c.switchMasterRetry(fmt.Sprintf("%s:%s", m[2], m[3]))
 				}
+			case "+slave", "+sdown", "-sdown":
+				m := strings.SplitN(event.Message, " ", 4)
+				if m[0] == "slave" {
```
Comment (rueian) [inline sentinel.go:343]: We might also need to check `c.readonly` here as well.

Comment (smf8) [reply sentinel.go:343]: I think beside `c.readonly` we need to check `c.sOpt.Sentinel.MasterSet` in these events. because other master set events can trigger this client's refresh with current implementation.

**Comment by smf8** [comment]
Hey @rueian, I've updated the tests to cover the changes in this PR. Let me know if I've missed anything.

File: sentinel_test.go [hunk 10]
```diff
@@ -162,6 +162,229 @@ func TestSentinelClientInit(t *testing.T) {
...
+							{typ: '+', string: "ip"}, {typ: '+', string: ""},
+							{typ: '+', string: "port"}, {typ: '+', string: "4"},
+						}},
+					}}},
+					{val: RedisMessage{typ: '*', values: []RedisMessage{
+						RedisMessage(*Nil),
+						//{typ: '%', values: []RedisMessage{
+						//	RedisMessage(RedisError{}),
+						//	RedisMessage(RedisError{}),
+						//}},
```
Comment (rueian) [inline sentinel_test.go:280]: Hi @smf8,

Just removing these comments and then I think this is good to merge. Thanks!

Comment (rueian) [reply sentinel_test.go]: Yes, they look good.
